inital version of spring boot rental notifier blog post (#17)
lunatech-labs/lunatech-blog: posts/2020-12-14-spring-boot-rental-notifier-bot.adoc

= Creating Telegram bot for rental notifications in Spring Boot
Daria Pindus <daria.pindus@lunatech.nl>
v1.0, 2020-12-14
:title: Creating Telegram bot for rental notifications in Spring Boot
:tags: [java, spring boot, telegram, bot]

> During my relocation to Rotterdam everything went smooth except one thing - rental process. There seemed to be not a lot of new offers, keeping track of interesting ones is a bit messy (we even created an Excel table, where we wrote down apartments and all details we were interested in). So using my passion for automation, to improve the onboarding process, and to refresh my knowledge on some technologies, I decided to create customizable Telegram bot to get notifications on new rental offers in a handy way.

Reviewer: ```suggestion
During my relocation to Rotterdam everything went smoothly except for one thing - the rental process. There seemed to be not a lot of new offers and keeping track of interesting ones was a bit messy (we even created an Excel table, where we wrote down apartments and all details we were interested in). So using my passion for automation, to improve the onboarding process, and to refresh my knowledge on some technologies, I decided to create a customizable Telegram bot to get notifications on new rental offers in a handy way.
```

You can find the source on https://github.com/DariaPindus/RealEstateHelper[GitHub], and you can also subscribe to the bot using @rental_notifier_bot

> Here how it looks like:

Reviewer: ```suggestion
Here is what it looks like:
```

So I decided to create a Telegram bot that would send me updates once interesting offers appear. I consider fetching data from only one website, but designing the application in a way that could support more data sources and new cities. During the development, I also thought about the ability of adding not only rental offers but real estate ones or other similar offers you want to be notified about, and as well to integrate chat bots for other messengers. Finally, my intention was to make it possible to automate the process of reply with a user-defined template.

Now let’s take a look at the User Stories Diagram that were completely or partially realized in this version:

image::../media/2020-12-14-spring-boot-rental-notifier-bot/use-case-diagram.png[Use cases, 600]

== Excursus: Why Telegram ?

Telegram is a pretty common messenger in Eastern European countries and their Bot API is a piece of cake. And I’m kinda a Telegram evangelist as it’s pretty cool and has a lot of great features you should take a look at (like secret chats, cloud storage, bots, groups and channels, voice and video calls, multiple sessions and so on).

== What technologies were used?

> I used following technologies:

Reviewer: ```suggestion
I used the following technologies:
```

Currently the user interacts with the application through the Telegram messenger. Another external dependency is the data source (Pararuis) where the data is extracted from.

And back to the technologies. My intention was to create an extensible server application that could be used for convenient real estate related notifications with possible analytic usage. Here is the component diagram for Rental Notifier:

image::../media/2020-12-14-spring-boot-rental-notifier-bot/component-diagram.png[Component Diagram, 700]

As I mentioned before, I aimed to design the application to be extensible for future needs.

I used a modular approach, creating 3 base modules (packages) - “bot”, “common”, “rental”. In case of an (unlikely) high growth of the application, such structure will allow splitting the monolith into nice simple microservices. As an additional benefit, the modules are easy to navigate, easy to extend with small risk to affect another module. Furthermore, the communication between them is done through the message broker - ActiveMQ. It facilitates adding independent, easy-to-scale and easy-to-extend modules by adding new topics and new listeners.

> As you can see, there are a few components. Following the best practices I tend to do them small, loosely-coupled, doing one thing at the time, open for extensions. As well I used abstractions and interfaces for easy updates - i.e. supporting new data sources (apart from Pararius) would be a question of few new classes implementing existing interfaces.

Reviewer: Sometimes in English "there are a few" can be confusing :) . Some people might interpret "there are a few components" to mean that "there are many components". But if I understand this paragraph you are trying to say that "there are NOT many components".

So you either need to say:
- "there are few components" (so delete the "a"), or explicitly say
- "there are not many components"

> As you can see, there are a few components. Following the best practices I tend to do them small, loosely-coupled, doing one thing at the time, open for extensions. As well I used abstractions and interfaces for easy updates - i.e. supporting new data sources (apart from Pararius) would be a question of few new classes implementing existing interfaces.

Reviewer: ```suggestion
As you can see, there are a few components. Following best practices I tend to make them small, loosely-coupled, doing one thing at the time, open for extension. As well, I used abstractions and interfaces for easy updates - i.e. supporting new data sources (apart from Pararius) would be a question of few new classes implementing existing interfaces.
```

Meanwhile, `RentalNotifierBot` itself can handle incoming messages. For executing all related logic there is `BotHandlerFacadeImpl` to help. This object has `BotReplier` dedicated to manipulating the user state and how to answer messages based on it. To make user state handling easy to extend I decided to create a map of state enum and its logic handler. To access the current state I use `UserCache`. For now, it’s purely in-memory storage - Java Map. But the idea is to use something efficient and fast as this action happen frequently. A better approach I think could be using Redis or something similar.

As the correct answer is created the message is formed and returned back to `RentalNotifierBot` that sends it to a user.

== So what is it all about?

The main feature and goal of the project was storing rental offers from different sources. It would allow them to be analyzed and notified about. The domain model allows to keep the history of offers changes, execution transactionality prevents data from being incorrectly/partially saved, logging allows keeping an eye on failures and mistakes.

> As it is a small pet project not everything intended has been done. Among possible enhancements I see following ones:

Reviewer: ```suggestion
As it is a small pet project not everything intended has been done. Among possible enhancements I see the following:
```

* Parallelizing the process of fetching data
* Better messages aggregation (sending them in batches and more failure-proof)
* Triggering the process of unsubscribing (by deleting conversation with bot) to clean user cache
* More customizations for offers
* API for analytics and better security for it

> _P.S._ By the time I ended doing this project and deploying it to the server we had already found an apartment, so I haven’t really used it. But as I keep being subscribed to it I kinda became an expert in Rotterdam real estates :)

Reviewer: ```suggestion
_P.S._ By the time I finished this project and deployed it to the server we had already found an apartment, so I haven’t really used it. But as I am still subscribed to it I kinda have become an expert in Rotterdam real estate :)
```
